pkg/controllers/mcs/mcs_controller.go

```diff
@@ -0,0 +1,382 @@
+	ctx context.Context,
+	mcs *networkingv1alpha1.MultiClusterService,
+	svc *corev1.Service,
+) (sets.Set[string], error) {
+	syncClusters := sets.New[string]()
+	clusters := &clusterv1alpha1.ClusterList{}
+	err := c.Client.List(ctx, clusters)
+	if err != nil {
+		klog.ErrorS(err, "failed to list clusters")
+		return syncClusters, err
+	}
+
+	serverLocations := sets.New[string](mcs.Spec.ServiceProvisionClusters...)
+	clientLocations := sets.New[string](mcs.Spec.ServiceConsumptionClusters...)
+	for _, cluster := range clusters.Items {
+		// if ServerLocations or ClientLocations are empty, we will sync work to the all clusters
+		if len(serverLocations) == 0 || len(clientLocations) == 0 ||
+			serverLocations.Has(cluster.Name) || clientLocations.Has(cluster.Name) {
+			syncClusters.Insert(cluster.Name)
+		}
+	}
+
+	svcObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(svc)
+	if err != nil {
+		return syncClusters, err
+	}
+
+	mcsID, err := c.getMultiClusterServiceID(mcs)
+	if err != nil {
+		klog.Errorf("Get MultiClusterService(%s/%s) ID error:%v", mcs.Namespace, mcs.Name, err)
+		return syncClusters, err
+	}
+
+	var errs []error
+	for clusterName := range syncClusters {
+		workMeta := metav1.ObjectMeta{
+			Name:       names.GenerateWorkName(svc.Kind, svc.Name, clusterName+"/"+svc.Namespace),
+			Namespace:  names.GenerateExecutionSpaceName(clusterName),
+			Finalizers: []string{util.ExecutionControllerFinalizer},
+			Labels: map[string]string{
+				networkingv1alpha1.MultiClusterServicePermanentIDLabel: mcsID,
+				util.ManagedByKarmadaLabel:                             util.ManagedByKarmadaLabelValue,
+			},
+		}
+
+		if err = helper.CreateOrUpdateWork(c, workMeta, &unstructured.Unstructured{Object: svcObj}); err != nil {
+			klog.Errorf("Failed to create or update resource(%v/%v) in the given member cluster %s, err is %v",
+				workMeta.GetNamespace(), workMeta.GetName(), clusterName, err)
+			c.EventRecorder.Event(mcs, corev1.EventTypeWarning, events.EventReasonSyncServiceWorkFailed, fmt.Sprintf(
+				"Failed to create or update resource(%v/%v) in member cluster(%s): %v",
+				workMeta.GetNamespace(), workMeta.GetName(), clusterName, err))
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return syncClusters, errors.NewAggregate(errs)
+	}
+
+	return syncClusters, nil
+}
+
+func (c *MCSController) updateMCSStatus(
+	ctx context.Context,
+	mcs *networkingv1alpha1.MultiClusterService,
+) error {
+	currentServiceAppliedCondition := meta.FindStatusCondition(mcs.Status.Conditions, networkingv1alpha1.MCSServiceAppliedConditionType)
+	if currentServiceAppliedCondition == nil {
+		mcs.Status.Conditions = append(mcs.Status.Conditions, metav1.Condition{
+			Message:            "Service is propagated to target clusters.",
+			Reason:             "ServiceAppliedSuccess",
+			Status:             metav1.ConditionTrue,
+			Type:               networkingv1alpha1.MCSServiceAppliedConditionType,
```

Comment:
I there there is only  `ServiceAppliedSuccess` state, can we also record the failed state?